src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -62,44 +62,62 @@ AsyncRPCOperation_sendmany::AsyncRPCOperation_sendmany(
-    if (sendFromAccount_ != ZCASH_LEGACY_ACCOUNT) {
-        allowedChangeTypes_.insert(libzcash::ChangeType::Sapling);
-    }
-
-    // calculate the target totals
+    // Determine the target totals and recipient pools
     for (const SendManyRecipient& recipient : recipients_) {
         std::visit(match {
             [&](const CKeyID& addr) {
                 transparentRecipients_ += 1;
                 txOutputAmounts_.t_outputs_total += recipient.amount;
-                allowedChangeTypes_.insert(libzcash::ChangeType::Transparent);
+                recipientPools_.insert(OutputPool::Transparent);
             },
             [&](const CScriptID& addr) {
                 transparentRecipients_ += 1;
                 txOutputAmounts_.t_outputs_total += recipient.amount;
-                allowedChangeTypes_.insert(libzcash::ChangeType::Transparent);
+                recipientPools_.insert(OutputPool::Transparent);
             },
             [&](const libzcash::SaplingPaymentAddress& addr) {
                 txOutputAmounts_.sapling_outputs_total += recipient.amount;
-                if (ztxoSelector_.SelectsSprout() && !allowRevealedAmounts_) {
-                    throw JSONRPCError(
+                recipientPools_.insert(OutputPool::Sapling);
+                if (!(ztxoSelector_.SelectsSapling() || allowRevealedAmounts_)) {
+                    if (ztxoSelector_.SelectsSprout()) {
+                        throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
+                            "Sending from the Sprout shielded pool to the Sapling "
+                            "shielded pool is not enabled by default because it will "
+                            "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                            "Resubmit with the `allowRevealedAmounts` parameter set to `true` if "
+                            "you wish to allow this transaction to proceed anyway.");
+                    }
+                    if (builder_.SupportsOrchard() && ztxoSelector_.SelectsOrchard()) {
+                        throw JSONRPCError(
                             RPC_INVALID_PARAMETER,
-                            "Sending between shielded pools is not enabled by default because it will "
+                            "Sending from the Orchard shielded pool to the Sapling "
+                            "shielded pool is not enabled by default because it will "
                             "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
                             "Resubmit with the `allowRevealedAmounts` parameter set to `true` if "
                             "you wish to allow this transaction to proceed anyway.");
+                    }
+                    // If the source selects transparent then we don't show an
+                    // error because we are necessarily revealing information.
```

Comment:
@ebfull asked about whether the `allowRevealing` flag should control transparent pool flows as well. I suggested that it should do so if the transparent coins are sourced from a Unified FVK or an account, or if there are transparent recipients from Unified Addresses, because the caller may not be aware that transparent coins are being selected.

In the case where the caller has specified a transparent address as the source (or `ANY_TADDR`), or has specified legacy transparent addresses as recipients, then the caller knows they are transparent and is already requesting the transaction be made. @nuttycom notes that there is still the potential for users being confused about what transparent addresses are, and I pointed out that gating `z_sendmany` calls that only involve legacy transparent addresses (not transparent UA receivers) would break compatibility with existing workflows. To compromise, we decided that if any source or recipient involves UAs (whether or not they have transparent components), then it's okay to have the flag gate the flow interactions of legacy addresses as well. So we think the only case that needs changes is legacy address UTXO being sent to UA receivers (because the caller might think we're sending to a shielded receiver when we're actually sending to a transparent receiver).

If in future we define a new RPC method that has clearer gating ("this is what your proposed transaction will reveal"), then that RPC can and should include legacy addresses in its gating.

Reply:
@ebfull noted that having a single `z_sendmany` flag for both pool-crossing and transparent-coin-using could be dangerous for users, in that a user that encounters the first case first might think "oh, I just set this flag and I can cross pool boundaries", and then blindly set that flag on every subsequent transaction. I noted that the flag's current name (which we've added for release in 4.7.0) is `allowRevealedAmounts`, which _can't_ cover transparent as that also reveals sender information. We agreed that we should add an `allowRevealedSenders` flag that allows transparent coins to be selected, and that if `allowRevealedSenders = true` then `allowRevealedAmounts` must also be set to true.

Reply:
Opened issues for the above: #5676, #5677, #5678, #5679.